File: Minimal/src/Brontide.Minimal.Binding/Portable/PortableProviderEndpoint.fs
namespace Brontide.Minimal.Binding.Portable

open System

/// What one provider Operation did.
///
/// A shaped failure is a case of the same union as a success because both are ordinary Outcomes:
/// the Operation reported structured details in its declared detail Shape, which is neither a
/// protocol nor a process failure and transports no exception.
[<StructuralEquality; NoComparison>]
type OperationEffect =
    | EffectSucceeded of result: PortableValue * providerEffectCount: int64
    | EffectFailed of details: PortableValue * providerEffectCount: int64

/// The provider's domain behind the binding.
///
/// This is where the provider's own admission happens for a cross-trust presentation: the handler
/// receives attributable context and exact addressing and decides for itself, reporting a refusal
/// as a shaped failed Outcome rather than as a protocol rejection.
type IPortableOperationHandler =
    abstract Invoke:
        operation: PortableOperationRef * input: PortableValue * resources: PortableResource list ->
            PortableResult<OperationEffect>

/// The provider's answer to one request, before any transport decision.
[<StructuralEquality; NoComparison>]
type ProviderOutcome =
    { Status: OutcomeStatus
      ValueShape: PortableShapeRef
      Value: PortableValue
      ProviderEffectCount: int64
      MappingObligations: string list
      Resources: ResourceObservation list
      CopyCount: int64 }

/// The semantic provider endpoint: lifecycle, negotiation, conformance, resource admission, and
/// dispatch, with no transport knowledge.
///
/// Both realizations drive this same endpoint, which is what makes their category-level
/// observations comparable: the direct call and the process seam differ in how a request arrives,
/// not in what the endpoint decides about it.
type PortableProviderEndpoint(offered: ContractDocument, handler: IPortableOperationHandler, realization: Realization) =

    let mutable lifecycle = Lifecycle.create false
    let mutable established: BindingPlan option = None

    /// Advances the lifecycle, or refuses the transition and leaves the previous state intact.
    let advance kind =
        Lifecycle.apply kind lifecycle
        |> Result.map (fun next -> lifecycle <- next)

    /// A refusal ends the binding, because the lifecycle's declared transition on a protocol error
    /// is to the failed state. Applying it here rather than in the transport is what keeps the
    /// direct realization's endpoint state equal to the process realization's.
    let failOnRefusal (result: PortableResult<'T>) =
        match result with
        | Error(Refused _) ->
            lifecycle <- Lifecycle.fail lifecycle
            result
        | Ok _
        | Error(Interrupted _) -> result

    let requirePlan () =
        match established with
        | Some plan -> Ok plan
        | None -> stateViolation "unestablished" "A request arrived before any contract was established."

    /// Attribution is never inferred from delivery, so a required Fragment must be present.
    let requireDeclaredFragments (value: PortableValue) (declaration: OperationDeclaration) =
        if List.isEmpty declaration.RequiredFragments then
            Ok()
        else
            let attached = PortableRecord.fragments value

            declaration.RequiredFragments
            |> iterate (fun fragment ->
                if Map.containsKey fragment attached then
                    Ok()
                else
                    invalidPayload
                        "required-fragment-absent"
                        $"Operation {PortableOperationRef.text declaration.Reference} requires Fragment {PortableFragmentRef.text fragment}.")

    member _.State = Lifecycle.state lifecycle

    member _.Plan = established

    member _.Realization = realization

    member _.Establish(required: ContractDocument, hostEndpoint: string) : PortableResult<EstablishAcceptedBody> =
        portable {
            do! advance EnvelopeKind.Establish

            let! plan =
                PortableNegotiation.negotiate
                    required
                    offered
                    realization
                    hostEndpoint
                    (PortableProviderRef.text offered.Provider)
                    "the provider endpoint offered the exact negotiated contract"

            lifecycle <- Lifecycle.declareReplayProtection (BindingPlan.replayProtectionDeclared plan) lifecycle
            do! advance EnvelopeKind.EstablishAccepted
            established <- Some plan

            return
                { Contract = offered
                  CompactIdentifiers = BindingPlan.compactAssignments plan }
        }
        |> failOnRefusal

    member _.SignalReady() = advance EnvelopeKind.Ready |> failOnRefusal

    member _.Withdraw() = advance EnvelopeKind.Withdraw |> failOnRefusal

    member _.Terminate() = advance EnvelopeKind.Terminate |> failOnRefusal

    member _.Fail() = lifecycle <- Lifecycle.fail lifecycle

    /// Resolves the Operation a request names, canonically or by a binding-scoped compact
    /// identifier. The transport needs the declaration before it can decode a schema-guided input.
    member _.ResolveOperation(designation: OperationDesignation) : PortableResult<OperationDeclaration> =
        portable {
            let! plan = requirePlan ()

            match designation with
            | OperationDesignation.Canonical reference -> return! BindingPlan.operation reference plan
            | OperationDesignation.Compact compact ->
                let assignment =
                    BindingPlan.compactAssignments plan
                    |> List.tryFind (fun candidate ->
                        candidate.Space = IdentitySpace.Operation
                        && CompactId.value candidate.Compact = compact)

                match assignment with
                | None ->
                    // A compact identifier this binding never assigned resolves to no canonical
                    // identity.
                    return!
                        unsupportedContract
                            "compact-identifier-unassigned"
                            $"Compact identifier {compact} was never assigned in this binding."
                | Some assignment ->
                    let reference =
                        BindingPlan.operations plan
                        |> List.tryFind (fun candidate -> PortableOperationRef.text candidate = assignment.Reference)

                    match reference with
                    | Some reference -> return! BindingPlan.operation reference plan
                    | None ->
                        return!
                            unsupportedContract
                                "compact-identifier-unassigned"
                                $"Compact identifier {compact} names no negotiated Operation."
        }

    member this.Request
        (
            request: ChannelRequestId,
            designation: OperationDesignation,
            inputShape: PortableShapeRef,
            input: PortableValue,
            resources: PortableResource list
        ) : PortableResult<ProviderOutcome> =
        portable {
            let! plan = requirePlan ()
            do! advance EnvelopeKind.Request
            let! next = Lifecycle.recordRequest request lifecycle
            lifecycle <- next
            let! declaration = this.ResolveOperation designation

            if BindingPlan.trustBoundaryCrossed plan then
                do! PortableAuthorityVocabulary.requireNoCapabilityValue input

            do!
                resources
                |> iterate (fun resource ->
                    let flavour = ResourceFlavor.token (PortableResource.flavor resource)

                    portable {
                        do!
                            ensure (List.contains flavour declaration.ResourceFlavors) (fun () ->
                                unsupportedContract
                                    "resource-flavor-unnegotiated"
                                    $"Operation {PortableOperationRef.text declaration.Reference} declares no resource flavor '{flavour}'.")

                        do!
                            ResourceCodec.admit
                                resource
                                (BindingPlan.resourceFlavors plan)
                                (BindingPlan.acceptedResourceHandles plan)
                                (BindingPlan.limits plan)
                    })

            let catalog = BindingPlan.catalog plan
            let! projected, obligations = PortableValueCodec.project catalog inputShape declaration.InputShape input
            do! PortableValueCodec.validate catalog declaration.InputShape declaration.RequiredFragments projected
            do! requireDeclaredFragments projected declaration

            let! effect =
                try
                    handler.Invoke(declaration.Reference, projected, resources)
                with _ ->
                    // The provider's own runtime failed. Only the portable category crosses: no
                    // exception, stack trace, or runtime type name is admitted into the observation
                    // or the frame.
                    refuse
                        ProtocolCategory.InternalProtocolFailure
                        "handler-failure"
                        "The endpoint cannot continue protocol processing."

            do! advance EnvelopeKind.Outcome

            let observations =
                resources |> List.map (fun resource -> PortableResource.observe resource realization)

            let status, valueShape, value, effectCount =
                match effect with
                | EffectSucceeded(result, count) -> OutcomeStatus.Succeeded, declaration.ResultShape, result, count
                | EffectFailed(details, count) -> OutcomeStatus.Failed, declaration.DetailShape, details, count

            return
                { Status = status
                  ValueShape = valueShape
                  Value = value
                  ProviderEffectCount = effectCount
                  MappingObligations = obligations
                  Resources = observations
                  CopyCount = observations |> List.sumBy (fun resource -> resource.Copies) }
        }
        |> failOnRefusal

File: Reference/src/Brontide.Reference.Experimental.Binding/Portable/PortableProviderEndpoint.cs
using System.Collections.Immutable;

namespace Brontide.Reference.Experimental.Binding.Portable;

/// <summary>What one provider Operation did.</summary>
public sealed record PortableOperationEffect(bool Succeeded, PortableValue Value, long ProviderEffectCount)
{
    public static PortableOperationEffect Success(PortableValue result, long effectCount) => new(true, result, effectCount);

    /// <summary>
    /// A shaped failed Outcome: the Operation reported structured failure details in its declared
    /// detail Shape. It is not a protocol or process failure and transports no exception.
    /// </summary>
    public static PortableOperationEffect Failure(PortableValue details, long effectCount) => new(false, details, effectCount);
}

/// <summary>
/// The provider's domain behind the binding.
/// </summary>
/// <remarks>
/// This is where the provider's own admission happens for a cross-trust presentation: the handler
/// receives attributable context and exact addressing and decides for itself, reporting a refusal
/// as a shaped failed Outcome rather than as a protocol rejection.
/// </remarks>
public interface IPortableOperationHandler
{
    PortableOperationEffect Invoke(
        PortableOperationReference operation,
        PortableValue input,
        IReadOnlyList<PortableResource> resources);
}

/// <summary>The provider's answer to one request, before any transport decision.</summary>
public sealed record PortableProviderOutcome(
    PortableOutcomeStatus Status,
    PortableShapeReference ValueShape,
    PortableValue Value,
    long ProviderEffectCount,
    ImmutableArray<string> MappingObligations,
    ImmutableArray<PortableResourceObservation> Resources,
    long CopyCount);

/// <summary>
/// The semantic provider endpoint: lifecycle, negotiation, conformance, resource admission, and
/// dispatch, with no transport knowledge.
/// </summary>
/// <remarks>
/// Both realizations drive this same endpoint, which is what makes their category-level
/// observations comparable: the direct call and the process seam differ in how a request arrives,
/// not in what the endpoint decides about it.
/// </remarks>
public sealed class PortableProviderEndpoint
{
    private readonly PortableContractDocument _offered;
    private readonly IPortableOperationHandler _handler;
    private readonly PortableRealization _realization;
    private readonly PortableLifecycle _lifecycle = new(replayProtectionDeclared: false);

    public PortableProviderEndpoint(
        PortableContractDocument offered,
        IPortableOperationHandler handler,
        PortableRealization realization)
    {
        _offered = offered ?? throw new ArgumentNullException(nameof(offered));
        _handler = handler ?? throw new ArgumentNullException(nameof(handler));
        _realization = realization;
    }

    public PortableBindingPlan? Plan { get; private set; }

    public PortableLifecycleState State => _lifecycle.State;

    public PortableEstablishAcceptedBody Establish(PortableContractDocument required, string hostEndpoint)
    {
        ArgumentNullException.ThrowIfNull(required);
        _lifecycle.Apply(PortableEnvelopeKind.Establish);
        var plan = PortableNegotiation.Negotiate(
            required,
            _offered,
            _realization,
            hostEndpoint,
            _offered.Provider.ToString(),
            "the provider endpoint offered the exact negotiated contract");
        _lifecycle.DeclareReplayProtection(plan.ReplayProtectionDeclared);
        _lifecycle.Apply(PortableEnvelopeKind.EstablishAccepted);
        Plan = plan;
        return new PortableEstablishAcceptedBody(_offered, plan.CompactIdentifierAssignments);
    }

    public void SignalReady() => _lifecycle.Apply(PortableEnvelopeKind.Ready);

    public void Withdraw() => _lifecycle.Apply(PortableEnvelopeKind.Withdraw);

    public void Terminate() => _lifecycle.Apply(PortableEnvelopeKind.Terminate);

    public void Fail() => _lifecycle.Fail();

    /// <summary>
    /// Handles one request, or refuses it with a portable category.
    /// </summary>
    /// <remarks>
    /// A refusal ends the binding, because the lifecycle's declared transition on a protocol error
    /// is to the failed state. Applying it here rather than in the transport is what keeps the
    /// direct realization's endpoint state equal to the process realization's.
    /// </remarks>
    public PortableProviderOutcome Request(
        PortableChannelRequestId requestId,
        PortableOperationReference? operation,
        int? compactOperation,
        PortableShapeReference inputShape,
        PortableValue input,
        IReadOnlyList<PortableResource> resources)
    {
        try
        {
            return RequestCore(requestId, operation, compactOperation, inputShape, input, resources);
        }
        catch (PortableFaultException)
        {
            _lifecycle.Fail();
            throw;
        }
    }

    private PortableProviderOutcome RequestCore(
        PortableChannelRequestId requestId,
        PortableOperationReference? operation,
        int? compactOperation,
        PortableShapeReference inputShape,
        PortableValue input,
        IReadOnlyList<PortableResource> resources)
    {
        ArgumentNullException.ThrowIfNull(input);
        ArgumentNullException.ThrowIfNull(resources);
        var plan = Plan ?? throw new PortableFaultException(
            PortableProtocolCategory.StateViolation,
            "unestablished",
            "A request arrived before any contract was established.");

        _lifecycle.Apply(PortableEnvelopeKind.Request);
        _lifecycle.RecordRequest(requestId);

        var declaration = ResolveOperation(operation, compactOperation);
        var reference = declaration.Reference;

        if (plan.TrustBoundaryCrossed)
        {
            PortableAuthorityVocabulary.RequireNoCapabilityContent(input);
        }

        foreach (var resource in resources)
        {
            if (!declaration.ResourceFlavors.Contains(resource.Flavor.Token(), StringComparer.Ordinal))
            {
                throw new PortableFaultException(
                    PortableProtocolCategory.UnsupportedContract,
                    "resource-flavor-unnegotiated",
                    $"Operation {reference} declares no resource flavor '{resource.Flavor.Token()}'.");
            }

            PortableResourceCodec.Admit(resource, plan.ResourceFlavors, plan.AcceptedResourceHandles, plan.Limits);
        }

        var projection = PortableValueCodec.Project(
            plan.Catalog,
            input,
            inputShape,
            declaration.InputShape,
            declaration.RequiredFragments);
        PortableValueCodec.Validate(
            plan.Catalog,
            declaration.InputShape,
            projection.Value,
            declaration.RequiredFragments);
        RequireDeclaredFragments(projection.Value, declaration);

        PortableOperationEffect effect;
        try
        {
            effect = _handler.Invoke(reference, projection.Value, resources);
        }
        catch (Exception exception) when (exception is not PortableFaultException and not PortableProcessFailureException)
        {
            // The provider's own runtime failed. Only the portable category crosses: no exception,
            // stack trace, or runtime type name is admitted into the observation or the frame.
            throw new PortableFaultException(
                PortableProtocolCategory.InternalProtocolFailure,
                "handler-failure",
                "The endpoint cannot continue protocol processing.");
        }

        _lifecycle.Apply(PortableEnvelopeKind.Outcome);

        var observations = resources
            .Select(resource => PortableResourceCodec.Observe(resource, _realization))
            .ToImmutableArray();
        return new PortableProviderOutcome(
            effect.Succeeded ? PortableOutcomeStatus.Succeeded : PortableOutcomeStatus.Failed,
            effect.Succeeded ? declaration.ResultShape : declaration.DetailShape,
            effect.Value,
            effect.ProviderEffectCount,
            projection.MappingObligations,
            observations,
            observations.Sum(resource => resource.Copies));
    }

    /// <summary>
    /// Resolves the Operation a request names, canonically or by a binding-scoped compact
    /// identifier. The transport needs the declaration before it can decode a schema-guided input.
    /// </summary>
    public PortableOperationDeclaration ResolveOperation(PortableOperationReference? operation, int? compactOperation)
    {
        var plan = Plan ?? throw new PortableFaultException(
            PortableProtocolCategory.StateViolation,
            "unestablished",
            "No contract is established, so no Operation can be resolved.");

        if (operation is { } reference)
        {
            return plan.Operation(reference);
        }

        var compact = compactOperation ?? throw PortableFaultException.Malformed(
            "operation-designation",
            "A request names its Operation either canonically or by one compact identifier.");
        var assignment = plan.CompactIdentifierAssignments.FirstOrDefault(candidate =>
            candidate.Space == PortableIdentitySpace.Operation && candidate.CompactId.Value == compact);
        if (assignment is null)
        {
            // A compact identifier this binding never assigned resolves to no canonical identity.
            throw new PortableFaultException(
                PortableProtocolCategory.UnsupportedContract,
                "compact-identifier-unassigned",
                $"Compact identifier {compact} was never assigned in this binding.");
        }

        return plan.Operation(plan.Operations.First(candidate => candidate.ToString() == assignment.Reference));
    }

    /// <summary>Attribution is never inferred from delivery, so a required Fragment must be present.</summary>
    private static void RequireDeclaredFragments(PortableValue value, PortableOperationDeclaration declaration)
    {
        if (declaration.RequiredFragments.IsEmpty)
        {
            return;
        }

        var attached = value is PortableRecordValue record
            ? record.Fragments.Keys.ToImmutableHashSet()
            : [];
        foreach (var fragment in declaration.RequiredFragments)
        {
            if (!attached.Contains(fragment))
            {
                throw PortableFaultException.InvalidPayload(
                    "required-fragment-absent",
                    $"Operation {declaration.Reference} requires Fragment {fragment}.");
            }
        }
    }
}
